src/invariant/OrderBookIsNotCrossed.cpp

```diff
@@ -0,0 +1,205 @@
+{
+    return double(offer.price.n) / double(offer.price.d);
+}
+
+static std::string
+checkCrossed(Asset const& a, Asset const& b, OrderBook const& orderBook)
+{
+    // if either side of order book for asset pair empty or does not yet exist,
+    // order book cannot be crossed
+    auto iterA = orderBook.find(a);
+    auto iterB = orderBook.find(b);
+    if (iterA == orderBook.end() || iterB == orderBook.end())
+    {
+        return {};
+    }
+
+    auto const& iterAB = orderBook.at(a).find(b);
+    auto const& iterBA = orderBook.at(b).find(a);
+    if (iterAB == (*iterA).second.end() || iterBA == (*iterB).second.end())
+    {
+        return {};
+    }
+
+    auto const& asks = orderBook.at(a).at(b);
+    auto const& bids = orderBook.at(b).at(a);
+    if (asks.empty() || bids.empty())
+    {
+        return {};
+    }
+
+    // check if crossed
+    auto lowestAsk = price(*asks.cbegin());
+    auto highestBid = 1 / price(*bids.cbegin());
+
+    // Allow highestBid to equal (not exceed) lowestAsk if at least one of the
+    // offers is passive.
+    if (lowestAsk == highestBid)
+    {
+        auto findEqualPriceNonPassive = [](auto const targetPrice,
+                                           auto const& offer) {
+            return price(offer) == targetPrice &&
+                   (offer.flags & PASSIVE_FLAG) == 0;
+        };
+
+        auto nonPassiveAskAtSamePrice =
+            std::find_if(asks.cbegin(), asks.cend(),
+                         std::bind(findEqualPriceNonPassive, lowestAsk,
+                                   std::placeholders::_1));
+        auto nonPassiveBidAtSamePrice =
+            std::find_if(bids.cbegin(), bids.cend(),
+                         std::bind(findEqualPriceNonPassive, 1 / lowestAsk,
+                                   std::placeholders::_1));
+
+        if (nonPassiveAskAtSamePrice != asks.cend() &&
+            nonPassiveBidAtSamePrice != bids.cend())
+        {
+            return fmt::format(
+                FMT_STRING("Order book is in a crossed state for {} - {} "
+                           "asset pair.\nEqual price {} for non-passive offers "
+                           "{} and {}."),
+                xdr::xdr_to_string(a), xdr::xdr_to_string(b), lowestAsk,
+                xdr::xdr_to_string(*nonPassiveBidAtSamePrice),
+                xdr::xdr_to_string(*nonPassiveAskAtSamePrice));
+        }
+
+        return {};
+    }
+
+    if (lowestAsk < highestBid)
+    {
+        return fmt::format(
+            FMT_STRING(
+                "Order book is in a crossed state for {} - {} "
+                "asset pair.\nTop of the book is:\n\tAsk price: {}\n\tBid "
+                "price: {}\n\nWhere {} >= {}!"),
+            xdr::xdr_to_string(a), xdr::xdr_to_string(b), lowestAsk, highestBid,
+            highestBid, lowestAsk);
+    }
+    return {};
+}
+
+void
+OrderBookIsNotCrossed::deleteFromOrderBook(OfferEntry const& oe)
+{
+    mOrderBook[oe.selling][oe.buying].erase(oe);
```

Comment:
Why isn't `mOrderbook` an `std::unordered_map<std::pair<Asset, Asset>, Orders>` ? All places I see that we're always looking things up with both assets fixed so we're just making things slower for no reason by nesting them.

Reply:
You're right -- that made several passes shorter.